changefeedccl: add webhook-chaos roachtest

asg0451 opened the pull request:
Add new webhook-chaos roachtest with order validation.

Epic: [CRDB-38755](https://cockroachlabs.atlassian.net/browse/CRDB-38755)
Part of: https://github.com/cockroachdb/cockroach/issues/124148

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126517)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: add webhook-chaos roachtest

Add new webhook-chaos roachtest with order validation.

Epic: CRDB-38755
Part of: #124148

Release note: none

DarrylWong commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -2563,7 +2602,7 @@ func (c *SyncedCluster) pgurls(
 		if err != nil {
 			return nil, err
 		}
-		m[node] = c.NodeURL(host, desc.Port, virtualClusterName, desc.ServiceMode, DefaultAuthMode(), "" /* database */)
```
Is there a reason this is being changed to `AuthUserCert`? `DefaultAuthMode` *should* default to AuthUserCert anyway.

rharding6373 reviewed (commented):
Very cool. Thanks for polishing this up!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O1rj-2vDZavZI2EDx6w:b-7wacm7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @DarrylWong, @renatolabs, and @wenyihu6)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 496 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1ZEmyXCqzOKGca5rK7:-O1ZEmyXCqzOKGca5rK8:bp8d0mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L496)):*
> ```Go
> 	sinkURIOverride string
> 
> 	// These options are only implemented for kafka and webhooks.
> ```

nit: s/These/sinkChaos and repeat the comment for validateOrder and any other relevant options.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rfc4L5ANTh7gEj9ro:-O1rfc4L5ANTh7gEj9rp:bmgzswg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L1550)):*
> ```Go
> 		Leases:           registry.MetamorphicLeases,
> 		CompatibleClouds: registry.AllExceptAWS,
> 		Suites:           registry.Suites(registry.Nightly),
> ```

Did you mean to include changes to the kafka chaos test in this PR?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rfnxUEMDN7pCQb3YF:-O1rfnxUEMDN7pCQb3YG:bfbjnox) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L1879)):*
> ```Go
> 		Leases:           registry.MetamorphicLeases,
> 		CompatibleClouds: registry.OnlyGCE,
> 		Suites:           registry.ManualOnly,
> ```

Why don't we want to run this test regularly?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3366 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rgwIf1aGwNDvD_CzH:-O1rgwIf1aGwNDvD_CzI:bi49z7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L3366)):*
> ```Go
> 		require.NoError(m.t, json.Unmarshal(scanner.Bytes(), &payload))
> 
> 		// NOTE: this does not validate resolved timestamps, since we don't know which topic they're for.
> ```

Could you clarify what you mean by "topic" here? I didn't think that webhook has the same notion of "topic" that kafka does. Are there actually different topics listed in the logs below in practice?


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, @renatolabs, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rfc4L5ANTh7gEj9ro:-O1rpsIk5g-ddduNkqy_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L1550)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Did you mean to include changes to the kafka chaos test in this PR?
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rfnxUEMDN7pCQb3YF:-O1rpzPt3WbwQz__rjKb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L1879)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Why don't we want to run this test regularly?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, @renatolabs, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3366 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1rgwIf1aGwNDvD_CzH:-O1s5qD6BCOqYeJ1ZbcT:b5kelim) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d75e88bad591e6cf14f7b0ef683a1edcc671f7/pkg/cmd/roachtest/tests/cdc.go#L3366)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you clarify what you mean by "topic" here? I didn't think that webhook has the same notion of "topic" that kafka does. Are there actually different topics listed in the logs below in practice?
</blockquote></details>

The webhook sink does have a notion of topic -- it sets a `topic` field in the payload. However, the resolved payloads don't have this. 

Thinking about this again though i think we can just `NoteResolved` this timestamp for each topic we're tracking


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> Reviewed 1 of 4 files at r1, 2 of 3 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @DarrylWong, @renatolabs, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/cluster.go` line 1984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1s32C00yY2Gs-5RmUQ:-O1s32C00yY2Gs-5RmUR:b-epfbm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8dafdd8208e6d03101cd91a04ba0a08c90675ec/pkg/cmd/roachtest/cluster.go#L1984)):*
> ```Go
> ) (stdout io.ReadCloser, stderr io.ReadCloser, errChan chan error, err error) {
> 	if ctx.Err() != nil {
> 		return nil, nil, nil, errors.Wrap(ctx.Err(), "cluster.Get")
> ```

s/cluster.Get/cluster.Tail
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1582 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sS0sx6i1BsJ799UHx:-O1sS0sx6i1BsJ799UHy:bngbnxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4db235a845259ec24de6382f3a1f23e2d78439d1/pkg/cmd/roachtest/tests/cdc.go#L1582)):*
> ```Go
> 		Leases:           registry.MetamorphicLeases,
> 		CompatibleClouds: registry.AllExceptAWS,
> 		// TODO(#122372): Add this to the nightly test suite after we fix the Kafka restart bug.
> ```

I think we probably want to put this comment back?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1884 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sSHlG2BJ_RsPus0g7:-O1sSHlG2BJ_RsPus0g8:b-kpdvx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L1884)):*
> ```Go
> 			err := c.Install(ctx, t.L(), ct.webhookSinkNode(), "go")
> 			if err != nil {
> 				t.Skip(err)
> ```

I'm a little worried this could lead to this test being silently skipped a lot. If it's truly rare, maybe we can just deal with closing out the issues that get auto-opened?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sHMA78dGO83w-02ap:-O1sHMA86gNOUFRaPdB2:b-r5g4xj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3173)):*
> ```Go
> 
> import (
> 	"log"
> ```

super nit: the imports should be alphabetized
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3204 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sKf5D74yvDvCLm9WZ:-O1sKf5D74yvDvCLm9W_:bmhyvt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3204)):*
> ```Go
> 	t                test.Test
> 	cluster          cluster.Cluster
> 	webhookSinkNodes option.NodeListOption
> ```

nit: it seems like this is always a single node, maybe it should be named `webhookSinkNode` (singular) instead?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sPerN5ioYO-F5ptG2:-O1sPerN5ioYO-F5ptG3:b-az7igh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3274)):*
> ```Go
> 			case <-m.doneCh:
> 				return nil
> 			case <-time.After(5 * time.Second):
> ```

Is there a reason for this additional sleep given that the `restartSink` channel is already signalled with a delay? Could we maybe document the reason with a comment?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3342 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sQjFB9Uk9VL2Kgvt6:-O1sQjFB9Uk9VL2Kgvt7:b-1wr6b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3342)):*
> ```Go
> 	validators := make(map[string]cdctest.Validator)
> 	scanner := bufio.NewScanner(stdout)
> 	buf := make([]byte, 0, 64*1024)
> ```

Could you maybe add a comment like `/* 1 MiB */` etc to document the size choice? Maybe this could also be a power of 2 using bitshifts?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3355 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sRNWG62Y7nQDdwGQs:-O1sRNWG62Y7nQDdwGQt:b5szjuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3355)):*
> ```Go
> 				resolved, err := hlc.ParseHLC(payload.Resolved)
> 				require.NoError(m.t, err)
> 				require.NoError(m.t, v.NoteResolved("0", resolved))
> ```

nit: maybe have `const partition = "0"` and use that throughout this func


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
Meta-comment (as someone not familiar with these tests): does validation need to happen "live"? Is there a reason we couldn't log everything we need and then do the validation on the logged events once we stop the changefeed? This would eliminate the need for something like `Tail`.

If we do need the live aspect of this, it would be nice if the function added wasn't so specific to the purposes of this test. The only thing the function really cares about is streaming the output of a command back to the test. We have most of the pieces for this, but no function currently exposes the ability to pass your own `io.Writer`. This would be a lot more reusable than a `Tail` function that runs a hardcoded `tail` command.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O1vngJbD5HZt25qt_-F:b-f14tgq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @DarrylWong, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/cluster.go` line 1990 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1voKLo736L_EbuZ7-K:-O1voKLo736L_EbuZ7-L:byio6z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/cluster.go#L1990)):*
> ```Go
> 	errChan = make(chan error, 1)
> 	stdoutRead, stdoutWrite := io.Pipe()
> 	go func() {
> ```

If this function panics, the entire roachtest process will crash.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1884 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sSHlG2BJ_RsPus0g7:-O1voaJt1W66yblyEgnT:b-hy2ddr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L1884)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

I'm a little worried this could lead to this test being silently skipped a lot. If it's truly rare, maybe we can just deal with closing out the issues that get auto-opened?
</blockquote></details>

This can be a regular `t.Fatal`. If `c.Install` fails, the test failure will automatically be marked as a flake and CDC won't be pinged. These failures are aggregated on a single issue marked with `X-infra-flake` (example: #124052).


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
Validation doesn't *need* to happen live, but the UX is a lot better if it does. Tests which fail chaos can fail in as few as 5 min, vs 30 min for the overall run.

You would prefer something like `RunAsync(ctx, cmd...) (stdout io.Writer, stderr io.Writer, asyncErrors chan error, err error)` ? I don't feel strongly either way tbh. What do you think @rharding6373 / @andyyang890 ?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O1wF5bc3fbvQ5TgMqg5:b-bp5p5p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, @renatolabs, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/cluster.go` line 1990 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1voKLo736L_EbuZ7-K:-O1wGFXF60qi3d4OwBM2:b-z6pecb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/cluster.go#L1990)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

If this function panics, the entire roachtest process will crash.
</blockquote></details>

True. 
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1884 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sSHlG2BJ_RsPus0g7:-O1wGl3g1Z-oBQhuGzde:b-4v270l) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L1884)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This can be a regular `t.Fatal`. If `c.Install` fails, the test failure will automatically be marked as a flake and CDC won't be pinged. These failures are aggregated on a single issue marked with `X-infra-flake` (example: #124052).
</blockquote></details>

I copied that from the other webhook roachtest but ok, fixed
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3204 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sKf5D74yvDvCLm9WZ:-O1wH8z77G46697Ue_YY:bu4xrnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3204)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: it seems like this is always a single node, maybe it should be named `webhookSinkNode` (singular) instead?
</blockquote></details>

It need not always be a single node, with the previous "multiple sink nodes" PR.


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, @renatolabs, @rharding6373, and @wenyihu6)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3204 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O1sKf5D74yvDvCLm9WZ:-O1wJsxLE838CUboiESP:b-wzquww) ([raw file](https://github.com/cockroachdb/cockroach/blob/4025d0c9326c944e88421ff262b1ecbca6b3a22d/pkg/cmd/roachtest/tests/cdc.go#L3204)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

It need not always be a single node, with the previous "multiple sink nodes" PR.
</blockquote></details>

nvm, fixed


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
Let's go with Renato's suggestion to expose the `io.Writer` if it's feasible. I'm not sure `RunAsync` is what was in mind (I can't find an existing function similar to that). Could you provide pointers and work with Renato on what might be the best way to make a more sustainable/reusable solution to tail?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O2jwF3MB2hc2UZD4Z6j:bgkakl8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @DarrylWong, @renatolabs, and @wenyihu6)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 494 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O2juXvGDuuOiQYNXZXB:-O2juXvGDuuOiQYNXZXC:bvywm6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7ee5f94f7406e7a4b03dea40c18e7f17a069504/pkg/cmd/roachtest/tests/cdc.go#L494)):*
> ```Go
> 	sinkURIOverride string
> 
> 	// Only implemented for kafka and webhooks.
> ```

nit: include `sinkChaos` in the comment
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 496 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-O2jubPNB2_FNc9o0CkX:-O2jubPNB2_FNc9o0CkY:b-4r3xxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7ee5f94f7406e7a4b03dea40c18e7f17a069504/pkg/cmd/roachtest/tests/cdc.go#L496)):*
> ```Go
> 	// Only implemented for kafka and webhooks.
> 	sinkChaos bool
> 	// Only implemented for kafka and webhooks.
> ```

nit: include `validateOrder` in the comment


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
How about a `SpawnE(..)` method? I prefer returning a new io.Reader to accepting a Writer, because it makes it more clear whose responsibility it is to close it. The running command should close its stdout/stderr

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O2z4WFF1R5hRbMuWLFk:b-a5xowo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, @renatolabs, and @wenyihu6)



<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
While I'm not necessarily against the direction you have so far, I think it would serve us better to _not_ introduce the new complexity and new `cluster` function and instead opt to allow the existing `Run` functions to receive overwrites for stdout and stderr.

Draft diff of what I'm thinking:

```diff
diff --git a/pkg/cmd/roachtest/cluster.go b/pkg/cmd/roachtest/cluster.go
index d53d99e56d2..7768a752c1d 100644
--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -2474,6 +2474,16 @@ func (c *clusterImpl) RunE(ctx context.Context, options install.RunOptions, args
 	}
 	defer l.Close()
 
+	// TODO(testeng): update roachprod.Run to read std{out,err} from
+	// options instead of passing them as arguments.
+	stdout, stderr := l.Stdout, l.Stderr
+	if options.Stdout != nil {
+		stdout = options.Stdout
+	}
+	if options.Stderr != nil {
+		stderr = options.Stderr
+	}
+
 	cmd := strings.Join(args, " ")
 	c.f.L().Printf("running cmd `%s` on nodes [%v]", roachprod.TruncateString(cmd, 30), nodes)
 	if c.l.File != nil {
@@ -2482,7 +2492,7 @@ func (c *clusterImpl) RunE(ctx context.Context, options install.RunOptions, args
 	l.Printf("> %s", cmd)
 	if err := roachprod.Run(
 		ctx, l, c.MakeNodes(nodes), "", "", c.IsSecure(),
-		l.Stdout, l.Stderr, args, options,
+		stdout, stderr, args, options,
 	); err != nil {
 		if err := ctx.Err(); err != nil {
 			l.Printf("(note: incoming context was canceled: %s)", err)
diff --git a/pkg/cmd/roachtest/tests/cdc.go b/pkg/cmd/roachtest/tests/cdc.go
index c3673f52ad9..ed50757a1cd 100644
--- a/pkg/cmd/roachtest/tests/cdc.go
+++ b/pkg/cmd/roachtest/tests/cdc.go
@@ -174,6 +174,16 @@ func (ct *cdcTester) startCRDBChaos() {
 		Env:     envVars,
 	}
 	ct.mon.Go(ch.Runner(ct.cluster, ct.t, ct.mon))
+
+	ct.mon.Go(func(ctx context.Context) error {
+		ct.webhookR, ct.webhookW = io.Pipe()
+
+		opts := option.WithNodes(sinkNode).
+			WithRetryDisabled().
+			WithStdout(ct.webhookW)
+
+		return ct.cluster.RunE(ctx, opts, "tail -F <some file>")
+	})
 }
 
 func (ct *cdcTester) setupSink(args feedArgs) string {
diff --git a/pkg/roachprod/install/run_options.go b/pkg/roachprod/install/run_options.go
index e85d5b86eed..dfb5d4b2a64 100644
--- a/pkg/roachprod/install/run_options.go
+++ b/pkg/roachprod/install/run_options.go
@@ -10,7 +10,11 @@
 
 package install
 
-import "github.com/cockroachdb/cockroach/pkg/util/retry"
+import (
+	"io"
+
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+)
 
 type RunOptions struct {
 	// RetryOptions are the retry options
@@ -28,6 +32,9 @@ type RunOptions struct {
 	// what the default behaviour is.
 	FailOption FailOption
 
+	Stdout io.Writer
+	Stderr io.Writer
+
 	// These are private to roachprod
 	Nodes       Nodes
 	Concurrency int
@@ -79,6 +86,16 @@ func (r RunOptions) WithRetryDisabled() RunOptions {
 	return r
 }
 
+func (r RunOptions) WithStdout(stdout io.Writer) RunOptions {
+	r.Stdout = stdout
+	return r
+}
+
+func (r RunOptions) WithStderr(stderr io.Writer) RunOptions {
+	r.Stderr = stderr
+	return r
+}
+
 func (r RunOptions) WithShouldRetryFn(fn func(*RunResultDetails) bool) RunOptions {
 	r.ShouldRetryFn = fn
 	return r
```

(I just updated an arbitrary location in `cdc.go` to illustrate how the new options would be used).

`SpawnE` introduces new background goroutines and also forces the caller to create another background goroutine to monitor for errors; these are tricky because, as mentioned earlier, an accidental panic on these goroutines leads to the entire nightly run crashing.

The approach suggested just requires a single `Go` function in the existing monitor, which would take care of canceling the context when any other background function returns an error.

Let me know if you have thoughts or if I'm missing something.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126517#-:-O2zZ3Gd1efig-gS-qvr:bp8gvub)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DarrylWong, and @wenyihu6)



<!-- Sent from Reviewable.io -->
